signer/src/wsts_state_machine.rs

```diff
@@ -30,33 +39,435 @@ use wsts::traits::Signer as _;
+            .enumerate()
+            .map(|(idx, key)| (idx as u32, key.into()))
+            .collect();
+
+        // The number of possible signers is capped at a number well below
+        // u32::MAX, so this conversion should always work.
+        let num_signers: u32 = signer_public_keys
+            .len()
+            .try_into()
+            .expect("the number of signers is greater than u32::MAX?");
+        let signer_key_ids = (0..num_signers)
+            .map(|signer_id| (signer_id, std::iter::once(signer_id + 1).collect()))
+            .collect();
+        let config = wsts::state_machine::coordinator::Config {
+            num_signers,
+            num_keys: num_signers,
+            threshold: threshold as u32,
+            dkg_threshold: num_signers,
+            message_private_key: message_private_key.into(),
+            dkg_public_timeout: None,
+            dkg_private_timeout: None,
+            dkg_end_timeout: None,
+            nonce_timeout: None,
+            sign_timeout: None,
+            signer_key_ids,
+            signer_public_keys,
+        };
+
+        let wsts_coordinator = fire::Coordinator::new(config);
+        Self(wsts_coordinator)
+    }
+
+    fn get_config(&self) -> Config {
+        self.0.get_config()
+    }
+
+    fn save(&self) -> SavedState {
+        self.0.save()
+    }
+
+    async fn load<S>(
+        storage: &S,
+        aggregate_key: PublicKeyXOnly,
+        signer_public_keys: impl IntoIterator<Item = PublicKey> + Send,
+        threshold: u16,
+        signer_private_key: PrivateKey,
+    ) -> Result<Self, error::Error>
+    where
+        S: storage::DbRead + Send + Sync,
+    {
+        let encrypted_shares = storage
+            .get_encrypted_dkg_shares(aggregate_key)
+            .await?
+            .ok_or(Error::MissingDkgShares(aggregate_key))?;
+
+        let public_dkg_shares: BTreeMap<u32, wsts::net::DkgPublicShares> =
+            BTreeMap::decode(encrypted_shares.public_shares.as_slice())?;
+        let party_polynomials = public_dkg_shares
+            .iter()
+            .flat_map(|(_, share)| share.comms.clone())
+            .collect::<Vec<(u32, PolyCommitment)>>();
+
+        let mut coordinator = Self::new(signer_public_keys, threshold, signer_private_key);
+
+        let aggregate_key = encrypted_shares.aggregate_key.into();
+        coordinator
+            .set_key_and_party_polynomials(aggregate_key, party_polynomials)
+            .map_err(Error::wsts_coordinator)?;
+        coordinator.current_dkg_id = 1;
+
+        coordinator
+            .move_to(WstsState::Idle)
+            .map_err(Error::wsts_coordinator)?;
+
+        Ok(coordinator)
+    }
+
+    fn process_message(
+        &mut self,
+        message: &Message,
+    ) -> Result<(Option<Packet>, Option<OperationResult>), Error> {
+        let packet = Packet::from_message(message);
+        self.0
+            .process_message(&packet)
+            .map_err(Error::wsts_coordinator)
+    }
```

Comment:
Isn't the trait default already enough for this?

Reply:
Yup, [76b9b49](https://github.com/stacks-network/sbtc/pull/1287/commits/76b9b4971028e57839d23c41324fdc000dae4dba)